cgp/local_search/evolution_strategies.py

```diff
@@ -0,0 +1,218 @@
+        mirrored_sampling: bool = True,
+        n_processes: int = 1
+    ) -> None:
+        """Evolution strategies using the natural gradient of multinormal
+        search distributions in natural coordinates.
+
+        Standard deviation of the search distribution for each
+        parameter is stored per individual. Offspring may use the
+        values of their parents if the number of used parameters has
+        not changed. Does not consider covariances between parameters.
+
+        Implementation and default values following Wierstra et
+        al. (2014). Natural evolution strategies. Journal of Machine
+        Learning Research, 15(1), 949-980.
+
+        Parameters
+        ----------
+        individual : Individual
+            Individual for which to perform local search.
+        objective : Callable[[Callable[[np.ndarray[float]], np.ndarray[float]]], float]
+            Objective function used for the local search. Needs to accept
+            an numpy-compatible function as first argument (e.g., from
+            IndividualBase.to_numpy()) and return a float representing the
+            fitness of the individual.
+        seed : int
+            Seed for internal random number generator.
+        learning_rate_mu : float, optional
+            Learning rate for mean of search distribution.
+            Defaults to 1.0.
+        learning_rate_sigma : float, optional
+            Learning rate for standard deviation of search
+            distribution. Defaults to (3 + log(dim)) / (5.0 *
+            sqrt(dim)), where dim is the dimension of the search space.
+        population_size : int, optional
+            Number of samples evaluated per generation. Defaults to 4
+            + floor(3 * log(dim))), where dim is the dimension of the
+            search space..
+        max_generations : int, optional
+            Maximal number of generations. Defaults to 10.
+        min_sigma : float, optional
+            Minimal value for standard deviation of search
+            distribution. Defaults to 1e-6.
+        fitness_shaping : bool, optional
+            Whether to use fitness shaping. Defaults to True.
+        mirrored_sampling : bool, optional
+            Whether to use mirrored sampling. WARNING: Doubles the number
+            of samples per generation. Defaults to True.
+        n_processes : int, optional
+            Number of parallel processes. Defaults to 1.
+
+        """
+        self.objective = objective
+        self.learning_rate_mu = learning_rate_mu
+        self.learning_rate_sigma = learning_rate_sigma
+        self.population_size = population_size
+        self.max_generations = max_generations
+        self.min_sigma = min_sigma
+        self.fitness_shaping = fitness_shaping
+        self.mirrored_sampling = mirrored_sampling
+        self.n_processes = n_processes
+
+        self.rng = np.random.RandomState(seed)
+
+        self.sigma: Dict[int, np.ndarray[float]] = {}
```

Comment:
Is this dictionary cleared at some point? Or do we just keep adding sigmas for the parameters of individuals even if these individuals are not longer evolved?

Reply:
it's not cleared. i'm generally not too happy with this handling of sigma. as @HenrikMettler pointed out above sigmas might be reused although the apply to different parameters if the offspring and parents have different but an identical number of active parameters. any suggestion for improvement is appreciated!